internal/telemetry/sensitivemetadataallowlist/redact.go

```diff
@@ -10,23 +14,36 @@ type redactMode int
 
 const (
 	redactAllSensitive redactMode = iota
-	// redactMarketing only redacts marketing-related fields.
-	redactMarketing
+	// redactMarketingAndUnallowedPrivateMetadataKeys only redacts marketing-related fields as well
+	// as unallowed private metadata keys.
+	redactMarketingAndUnallowedPrivateMetadataKeys
 	// redactNothing is only used in dotocm mode.
 	redactNothing
 )
 
 // 🚨 SECURITY: Be very careful with the redaction mechanisms here, as it impacts
 // what data we export from customer Sourcegraph instances.
-func redactEvent(event *telemetrygatewayv1.Event, mode redactMode) {
-	// redactNothing
+func redactEvent(event *telemetrygatewayv1.Event, mode redactMode, allowedPrivateMetadataKeys []string) {
+	// redactNothing (generally in dotcom)
 	if mode >= redactNothing {
 		return
 	}
 
-	// redactMarketing
+	// redactMarketingAndUnallowedPrivateMetadataKeys
 	event.MarketingTracking = nil
-	if mode >= redactMarketing {
+	if mode >= redactMarketingAndUnallowedPrivateMetadataKeys {
+		// Do private metadata redaction in this if case, as the next case will strip
+		// everything
+		for key, value := range event.Parameters.PrivateMetadata.Fields {
+			if !slices.Contains(allowedPrivateMetadataKeys, key) {
+				// Strip all keys that are NOT in the allowlist
+				delete(event.Parameters.PrivateMetadata.Fields, key)
+			} else if _, isString := value.Kind.(*structpb.Value_StringValue); !isString {
+				// Strip all values that are not strings, even if they ARE in the allowlist
+				event.Parameters.PrivateMetadata.Fields[key] =
+					structpb.NewStringValue("ERROR: value of allowlisted key was not a string")
```

Comment:
This will emit the actual type we got in the error value:

```suggestion
					structpb.NewStringValue(fmt.Sprintf("ERROR: value of allowlisted key was not a string, got: %T", value.Kind))
```

Reply:
Great suggestion, added!